experimenter/experimenter/nimbus_ui_new/forms.py

```diff
@@ -389,6 +393,258 @@ def get_changelog_message(self):
+            "hx-post": reverse(
+                "nimbus-new-partial-update-branches", kwargs={"slug": self.instance.slug}
+            ),
+            "hx-trigger": "change",
+            "hx-select": "#branches",
+            "hx-target": "#branches",
+        }
+        self.fields["is_rollout"].widget.attrs.update(update_on_change_attrs)
+        self.fields["feature_configs"].widget.attrs.update(update_on_change_attrs)
+        self.fields["equal_branch_ratio"].widget.attrs.update(update_on_change_attrs)
+        self.fields["is_localized"].widget.attrs.update(
+            {
+                **update_on_change_attrs,
+                "hx-select": "#localization",
+                "hx-target": "#localization",
+            }
+        )
+
+    @property
+    def errors(self):
+        errors = super().errors
+        if any(self.branches.errors):
+            errors["branches"] = self.branches.errors
+        return errors
+
+    def save(self, *args, **kwargs):
+        experiment = super().save(*args, **kwargs)
+        self.branches.save()
+
+        if experiment.is_rollout:
+            branches = experiment.branches.all()
+
+            if experiment.reference_branch:
+                branches = branches.exclude(id=experiment.reference_branch.id)
+
+            branches.delete()
+
+        saved_experiment_feature_configs = set(experiment.feature_configs.all())
+        saved_branch_feature_configs = {
+            fv.feature_config
+            for b in experiment.branches.all()
+            for fv in b.feature_values.all()
+        }
+        new_feature_configs = (
+            saved_experiment_feature_configs - saved_branch_feature_configs
+        )
+        deleted_feature_configs = (
+            saved_branch_feature_configs - saved_experiment_feature_configs
+        )
+
+        for branch in experiment.branches.all():
+            branch.feature_values.filter(
+                feature_config__in=deleted_feature_configs
+            ).delete()
+
+            for feature_config in new_feature_configs:
+                branch.feature_values.create(feature_config=feature_config)
+
+        if experiment.equal_branch_ratio:
+            experiment.branches.all().update(ratio=1)
+
+        return experiment
+
+    def get_changelog_message(self):
+        return f"{self.request.user} updated branches"
+
+
+class NimbusBranchCreateForm(NimbusChangeLogFormMixin, forms.ModelForm):
+    class Meta:
+        model = NimbusExperiment
+        fields = []
+
+    def save(self, *args, **kwargs):
+        experiment = super().save(*args, **kwargs)
+        if not experiment.reference_branch:
+            name = "Control"
+        else:
+            branch_names_all = [f"Treatment {chr(i)}" for i in range(65, 91)]
+            branch_names_used = experiment.branches.values_list("name", flat=True)
+            branch_names_free = sorted(set(branch_names_all) - set(branch_names_used))
+            name = branch_names_free[0]
```

Comment:
We are generating the entire list of 26 branch names each time. What about a generator instead?

```suggestion
            branch_names_all = (f"Treatment {chr(i)}" for i in range(65, 91))
            branch_names_used = experiment.branches.values_list("name", flat=True)
            name = next(
                 branch_name
                 for branch_name in branch_names_all
                 if branch_name not in branch_names_used
            )
```

Reply:
Yeah theoretically, but in general I wouldn't use a generator unless it's either

1) Some very large dataset where precomputing it is actually costly
2) Involves IO

Here it's a fixed list of 26 strings which is so insignificantly small I don't think it makes a difference.

Reply:
Except it's not a fixed list of 26 strings, it's a list comprehension.

Reply:
Yeah but the resulting list is of fixed size is my point